packages/thyra/adapters/epetra/test/TestThyraDebugHang.cpp

```diff
@@ -53,45 +61,123 @@ TEUCHOS_UNIT_TEST( ThyraEpetraMultiVector, HangingInParallelDebug )
      os << prefix << "Creating Teuchos::Comm" << endl;
      cerr << os.str ();
    }
-   Teuchos::RCP<const Teuchos::Comm<Teuchos_Ordinal> > comm =
+   RCP<const Teuchos::Comm<Teuchos_Ordinal> > comm =
      Teuchos::DefaultComm<Teuchos_Ordinal>::getComm ();
+   // Make sure that everything is OK on all processes.
+   TEST_ASSERT( ! comm.is_null () );
+   lclSuccess = success ? 1 : 0;
+   reduceAll (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+   TEST_EQUALITY( gblSuccess, 1 );
+   if (gblSuccess != 1) {
+     out << "FAILED; some process(es) have a null Teuchos::Comm" << endl;
+     return;
+   }
 
    // Some processors have to have data and some not.
    const int localDim  = myRank % 2;
    const int globalDim = numProcs / 2;
-   Teuchos::RCP<const Epetra_Map> epetra_map;
+   RCP<const Epetra_Map> epetra_map;
    {
      std::ostringstream os;
      os << prefix << "Creating Epetra_Map: localDim=" << localDim << ", globalDim=" << globalDim << endl;
      cerr << os.str ();
    }
-   epetra_map = Teuchos::rcp(new Epetra_Map(globalDim,localDim,0,epetra_comm));
+   epetra_map = rcp (new Epetra_Map (globalDim, localDim, 0, epetra_comm));
+   // Make sure that everything is OK on all processes.
+   TEST_ASSERT( ! epetra_map.is_null () );
+   lclSuccess = success ? 1 : 0;
+   reduceAll (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+   TEST_EQUALITY( gblSuccess, 1 );
+   if (gblSuccess != 1) {
+     out << "FAILED; some process(es) have a null Epetra_Map" << endl;
+     return;
+   }
+
    {
      std::ostringstream os;
      os << prefix << "Creating Thyra::DefaultSpmdVectorSpace" << endl;
      cerr << os.str ();
    }
-   Teuchos::RCP<Thyra::DefaultSpmdVectorSpace<double>> SPMD = Thyra::DefaultSpmdVectorSpace<double>::create();
+   RCP<Thyra::DefaultSpmdVectorSpace<double> > SPMD =
+     Thyra::DefaultSpmdVectorSpace<double>::create();
+
+   // Make sure that everything is OK on all processes.
+   TEST_ASSERT( ! epetra_map.is_null () );
+   lclSuccess = success ? 1 : 0;
+   reduceAll (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+   TEST_EQUALITY( gblSuccess, 1 );
```

Comment:
Seems like we need to add some helper macros for this idiom like `TEST_ASSSERT_ON_ALL_PROCS( comm, localTestVal)`.  That would compress this test code a good bit.  But that can be a todo for later to add to Teuchos.